### DISCUSSION TITLE
Setting Relationship Parameter as a Function

### DISCUSSION THREAD

**Discussion by liamfennelly** [spine-tools/SpineOpt.jl#888] [Support]
Hello, I am relatively new to Spine, and I am having an issue with the relationship parameter editor to define a connection constraint for a unit_node_node relationship. 

One of the nodes defines household electricity demand for the year. The other node is an EV battery that stores energy. The unit is the EV charger which can both supply electricity to the EV from the household, and also from the household to the EV. The model optimises the power flows to minimise electricity costs by deciding when to charge the EV battery and when to discharge the battery to power the home.

I want to set the efficiency of the charger in both directions to be a function of the power flow rate at any given time. The formula that determines the charger efficiency at a certain power flow is as follows:

Efficiency = -9*( P_flow ^4) + 22*( P_flow ^3) - 18*( P_flow ^2) + 6*( P_flow)

In the relationship parameter editor I am only able to set the efficiency of the charger as a set value or a time series. However, I do not know what the efficiency will be at any given time as it depends on what the model sets the power flow to be. Is there a way for me to define this efficiency constraint as a function of the power flow rate and if so how?

**Comment by tarskul** [comment]
To get started with efficiencies in SpineOpt, we refer to this part of the documentation: [How to define an efficiency](https://spine-tools.github.io/SpineOpt.jl/latest/how_to/define_an_efficiency/)

However, it seems that your problem is highly non-linear which is not supported by SpineOpt. If you can accept the loss of accuracy, when approximating the non-linear efficiency by a piecewise-linear efficiency, you can use SpineOpt. The parameter that allows for piecewise linear efficiencies should be the `Incremental_heat_rate`.

**Reply by liamfennelly** [reply]
Hello, thank you, this is a useful way to approximate. However, Ive developed another method to approximate the efficiency of the charger in this example. It involves using a fixed efficiency of 90% as well as having a fixed loss of 0.15 kW. Therefore the output flow from the charger can be calculated as:

0.9*P_Flow - 0.15

Is there a way to put in this fixed loss of 0.15 kW. Will it be necessary to create another node connected to the charger as a power dump for this? The problem with this is that it is unknown at what times the charger will be operating, and therefore what times the power dump should be operating. Is it possible to set this fixed loss within a unit_node_node parmater?

**Comment by DillonJ** [comment]
@liamfennelly if using `incremental_heat_rate`, then the [unit_idle_heat_rate](https://spine-tools.github.io/SpineOpt.jl/latest/concept_reference/Parameters/#unit_idle_heat_rate) is a fixed loss.
